File: NeetCode_Roadmap/combination_sum.py
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        result = []

        def backtrack(start: int, remaining: int, cur: List[int]):
            if remaining == 0:
                result.append(cur.copy())
                return
            if remaining < 0:
                return
            
            for i in range(start, len(candidates)):
                candidate = candidates[i]

                cur.append(candidate)
                backtrack(i, remaining - candidate, cur)
                cur.pop()

        backtrack(0, target, [])

        return result

File: NeetCode_Roadmap/combination_sum.cpp
class Solution {
public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<vector<int>> result;
        vector<int> empty = {};
        backtrack(0, 0, target, candidates, empty, result);
        return result;
    }

    void backtrack(
        int idx, int sum, int target, vector<int>& candidates, vector<int>& cur_comb, vector<vector<int>>& result)
    {
        if (sum == target)
        {
            result.emplace_back(cur_comb);
            return;
        }

        if (idx >= candidates.size() || sum > target)
            return;

        cur_comb.push_back(candidates[idx]);
        backtrack(idx, sum+candidates[idx], target, candidates, cur_comb, result);
        cur_comb.pop_back();
        backtrack(idx+1, sum, target, candidates, cur_comb, result);
    }
};